nixos/modules/system/activation/activatable-system.nix

```diff
@@ -0,0 +1,48 @@
+/*
+  This module adds the activation script to toplevel, so that any previously
+  built configuration can be activated again, as long as they're available in
+  the store, e.g. through the profile's older generations.
+
+  Alternate applications of the NixOS modules may omit this module, e.g. to
+  build images that are pre-activated and omit the activation script and its
+  dependencies.
+ */
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib)
+    optionalString
+    ;
+in
+{
+  config = {
+    system.systemBuilderArgs = {
+      activationScript = config.system.activationScripts.script;
+      dryActivationScript = config.system.dryActivationScript;
+      installBootLoader = config.system.build.installBootLoader;
+      localeArchive = "${config.i18n.glibcLocales}/lib/locale/locale-archive";
+      distroId = config.system.nixos.distroId;
+      perl = pkgs.perl.withPackages (p: with p; [ ConfigIniFiles FileSlurp ]);
+    };
+
+    system.systemBuilderCommands = ''
+      echo "$activationScript" > $out/activate
+      echo "$dryActivationScript" > $out/dry-activate
+      substituteInPlace $out/activate --subst-var out
+      substituteInPlace $out/dry-activate --subst-var out
+      chmod u+x $out/activate $out/dry-activate
+      unset activationScript dryActivationScript
+
+      mkdir $out/bin
+      substituteAll ${./switch-to-configuration.pl} $out/bin/switch-to-configuration
+      chmod +x $out/bin/switch-to-configuration
+      ${optionalString (pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform) ''
```

Comment:
Don't you only need to have `perl` available on the `hostPlatform` ?

Reply:
(or on host and build?)

Reply:
`@perl@` is substituted in `switch-to-configuration.pl`.
`substituteAll` obscures that. Would be nice to be explicit about the variables.

Reply:
Apologies, my comment was about whether you needed to guard on same host platform and build platform to run the Perl check and I was wondering as Perl is like just checking a script if it could run beyond that condition.

Reply:
I suppose you could bring a separate perl for that purpose, but I doubt that it'd be worth the effort.

Reply:
That's fair